### PR TITLE
Add signature method for a scale. Closes #89

### LINKED ISSUE #89: Key signatures
Sorry for asking maybe some obvious question, but it’s not clear for me, what is the best way to get key signature notes for a scale?
I could imagine some method like `scale.signature()`, which returns key signature notes in proper order, but seems it is already possible to do, but it’s not quite clear how.
Any ideas?
Thank you.

### PR DESCRIPTION
**Pull request by dfcreative**
Hi, @matthewconstantine!
I’ve implemented, as it seems to me so far, the most logical method of getting key signatures for any scale with any tonic, based on note coordinates. It is quite late here and there might be some mistake, but tests of basic major/minor scales are passed. So please take a look, it should be nice ).
It closes #89.

### CODE REVIEW

**Comment by saebekassebil** [comment]
Hey Deema - this looks like a really good addition to the library! Let me just look it through and ponder it a bit, and I'll get it merged!

Cheers!

**Comment by dfcreative** [comment]
@saebekassebil so what thoughts on that?

**Comment by dfcreative** [comment]
@saebekassebil I don’t want to bother you, but do you have any considerations?

File: README.md [hunk 1]
```diff
@@ -181,6 +181,8 @@ The chroma number is ranging from pitch class C which is 0 to 11 which is B
 #### Note#interval(interval)
  - A sugar function for calling teoria.interval(note, interval);
 
+Returns a new note based on the passed interval.
```
Comment (saebekassebil) [inline README.md]: This comment isn't related the the PR. Please remove it.

File: README.md [hunk 2]
```diff
@@ -415,6 +417,9 @@ scale step. E.g. 'first', 'second', 'fourth', 'seventh'.
 
 *showOctaves* - A boolean meaning the same as `showOctaves` in `Note#solfege`
 
+#### Scale#signature()
+ - Returns the sequence of notes representing the key signature of a scale
```
Comment (saebekassebil) [inline README.md]: If you could add an example of how to use it, and which values the method returns, that'd be great!

File: lib/scale.js [hunk 3]
```diff
@@ -102,6 +102,32 @@ Scale.prototype = {
...
+   */
+  signature: function() {
+    var sequence = this.notes().sort(function (a, b) {
+      var aCoord = a.coord[1] < 0 ? a.coord[1] + 12 : a.coord[1];
+      var bCoord = b.coord[1] < 0 ? b.coord[1] + 12 : b.coord[1];
+      return aCoord - bCoord;
+    });
+
+    //if at least one flat - return flats signature
+    if (sequence.some(function (note) {
```
Comment (saebekassebil) [inline lib/scale.js]: If you could maybe break this into two lines (a bit more descriptive):

    var hasFlat = sequence.some(function (note) {
      return note.accidentalValue() < 0;
    });
    
    if (hasFlat) { ... }

**Comment by saebekassebil** [comment]
There might be a problem with this approach. I guess you'd maybe want to use this for drawing the signature for sheet music. If that's what you wanted, then you'd want an abstract "f#" and not maybe "f#3" or "f#8".

This method will return different results for:
```js
teoria.note('A8').scale('major').signature() != teoria.note('A3').scale('major').signature()
```

What are your thoughts on that?

**Comment by dfcreative** [comment]
Yes, that’s a good point.
Signatures returned should bring abstract meaning, but as far each specific note has that info about it’s abstract form - that’s ok.
However, for sheet music it is probably more handy to return not arbitrary formed key sequence (as it is done now), but a sequence corresponding to a clef. E.g. for the G-clef it’d be handy to have `teoria.note('A1').scale('major').signature('treble').toString(); //f#4,c#4,g#4`.
Description of all the clefs can be found [on the wikipedia](https://en.wikipedia.org/wiki/Clef).
So probably it might be a solution - to pass exact clef which it is needed to embody returned key sequence?
```js
//default clef is 'G', or 'treble'
teoria.note('A1').scale('major').signature().toString(); //f#4,c#4,g#4

//other clef can be passed as a param, e.g. 'bass', or 'F'
teoria.note('A1').scale('major').signature('F').toString(); //f#3,c#3,g#3
```
What do you think?

**Comment by saebekassebil** [comment]
When we begin to discuss "clefs" and such I think we're slowly progressing into sheet notation area, and I don't think this library should be coupled to sheet notation specifically - that's for another module.

I think maybe what we want is a method that returns the "position" of the signature/scale in the circle of fifths. E.g.:
```js
teoria.note('A').scale('major').signature() === 3
teoria.note('C').scale('major').signature() === 0
teoria.note('D').scale('minor').signature() === -2
// etc
```

and then we'd want another module that would transform this position into the notes in a given clef position:
```js
var getNotes = require('signature-for-clef');
var getTrebleNotes = getNotes('treble');
var signature = teoria.note('A').scale('major').signature(); // -> 3
getTrebleNotes(signature); // -> ["f#4", "c#4", "g#4"]
getNotes('bass')(signature) // -> ["f#3", "c#3", "g#3"]
```

What are your thoughts on this?

**Comment by dfcreative** [comment]
I see, you want to implement the principle of atomizing and functional separation in here. That makes perfect sense, but I felt teoria is rather a somewhat god-object, alike to jquery, including everything possible here at place to work with the music theory. (and it is hardly to find anything bad in that, as the final bundle size is bearable and relatively small). But in case of separation on plugins it ceases to be like that and becomes similar to [tonal](danigb/tonal), which was initially inspired by teoria but has basically plugins-composable architecture.
I can implement either approach, whether it is a separate module or teoria-function, but this is architectural decision, and it is up to you)

**Comment by dfcreative** [comment]
By the way, considering the [SMUFL](http://www.smufl.org/), what else can be possibly here to complicate the teoria? I hardly can think up of something.

**Comment by saebekassebil** [comment]
[tonal](https://github.com/danigb/tonal) is definitely a step in the right direction! What I'd like to see is that `teoria` could continue to be a "god-object" or a "programming platform for music theory". But it should consist of many smaller modules, like `tonal`, `pitch`, etc. The difference is, that I've been representing notes in a numeric format, while `tonal` prefers it to be in `string` format.

I'm not sure how you think SMUFL fits in?

**Comment by saebekassebil** [comment]
Hi Dima,

I just wanted to know if you're still interested in implementing this feature - have you come up with any good implementation ideas this year? ;)

I think we should go with the modular approach. But maybe including the module in teoria, as to remain "god" status?

**Comment by dfcreative** [comment]
Hi @saebekassebil!
We have chosen the similar approach in [audiojs/audio](https://github.com/audiojs/audio/tree/class) — it is going to be a class consisting of small components, but still monolithic thing, because for high-level apps that makes more sense.
I am a bit out of this issue though.

**Comment by dfcreative** [comment]
@saebekassebil is there any chance to have this PR merged? If not I suggest closing it then.

**Comment by saebekassebil** [comment]
Hey @dfcreative Thank you for the time invested in this pull - sorry for not being able to return back sooner.

I've closed the pull request since we never came to any conclusion about where to draw the line between sheet music and music theory - and also since you've implemented them in your own module, this seems redundant.

**Comment by dy** [comment]
@saebekassebil yes, thanks for making decision. It is fine - just wanted to close old unfinished things. I will get back to it if I need it one more time.

**Pull request closed without merging**